ServantPersistent/src/Operation.hs

```diff
@@ -1,70 +1,105 @@
 {-# LANGUAGE UndecidableInstances #-}
 {-# LANGUAGE MultiParamTypeClasses #-}
 {-# LANGUAGE GeneralizedNewtypeDeriving #-}
 module Operation 
   ( OperationT
   , Permission(..)
+  , PermissionError(..)
   , requirePermission
   , runOperation
+  , unsafeRunOperation
     ) where
+
 import Prelude hiding (null, filter)
 import Data.Set
-import Data.Text (Text)
 import Control.Monad.Trans
 import Control.Monad.Reader
-import Control.Monad.State
+import Control.Monad.State.Strict
 import Control.Monad.Except
 import Control.Applicative
-import Control.Monad.Writer.Strict
+import Control.Monad.Identity
 import qualified Database.Persist as DB
-import Control.Lens
+import Control.Lens ((^.))
 import Types
 import DBTypes
 import Models
 
-data Permission = EditUser UserID
-                | EditTenant TenantID
-                | ViewUser UserID
+data Permission = EditUser UserId
+                | EditTenant TenantId
+                | ViewUser UserId
+                | AssignRole UserId
+                | EditProduct ProductId
+                | CreateProduct TenantId
                     deriving (Eq,Ord,Show)
 
 data PermissionError = Requires (Set Permission)
 
-newtype OperationT m a = Op { unsafeRunOp :: WriterT (Set Permission) m a }
-    deriving (Functor, Applicative, Monad, Alternative, MonadPlus, Foldable, MonadFix, MonadTrans, MonadIO)
+newtype OperationT m a = Op { unsafeRunOp :: StateT (Set Permission) m a }
+    deriving (Functor, Applicative, Monad, Alternative, MonadPlus, MonadFix, MonadTrans, MonadIO)
+
 
 deriving instance (MonadReader r m) => MonadReader r (OperationT m)
-deriving instance (MonadState s m) => MonadState s (OperationT m)
+--deriving instance Monad m => MonadState (Set Permission) (OperationT m)
 deriving instance (MonadError e m) => MonadError e (OperationT m)
 
-requirePermission :: Monad m => Permission -> OperationT m ()
-requirePermission = Op . tell . singleton
+requirePermission :: Monad m => Permission -> OperationT m a -> OperationT m a
+requirePermission p o = (Op $ modify (insert p) ) >> o
+
+modifyM
+  :: (MonadState s (t m), MonadTrans t, Monad m) =>
+     (s -> m s) -> t m ()
+modifyM f = get >>= lift . f >>= put
 
-runOperation :: OperationT App a -> User -> ExceptT PermissionError App a
-runOperation op u@(UserB{_userRole=role}) = do
-    (a,s) <- lift $ runWriterT $ unsafeRunOp op
-    let go s (EditUserDetails) = 
-            fromList <$> 
-              filterM (\case (EditUser uid) -> hasTenant uid (_userTenantID u)
-                             _ -> return False)
-                             (toList s)
-        go s (EditTenantDetails) = 
-            fromList <$> 
-              filterM (\case (EditTenant tid) -> return $ tid == (_userTenantID u)
-                             _ -> return False)
-                             (toList s)
-        go s _ = return s
-    s' <- lift $ foldM go s (roleCapabilities role)
-    if null s'
+runOperation :: (MonadIO m)
```

Comment:
`runOperation` basically takes a permission protected operation, and a user, and returns the result of the operation only if the user has the required permissions to run the operation. Otherwise it throws an error describing which permissions are necessary. 

It does so by first generating the set of permissions required to perform the operation. Then by iterating over the capabilities of the user, it removes the permissions that the user has. If any permissions are leftover, it returns an error. Otherwise, it returns the result of the operation.

This ensures that the server never has access to the results of db API call that the currently logged on user can't access.